### ISSUE TITLE
Very high CPU usage vs. serenity::voice module

### ISSUE THREAD

**Issue by JellyWX** [serenity-rs/songbird#42]
I'm seeing very high CPU climb from Songbird vs. using the old `serenity::voice` module. Where `serenity::voice` sits at 0.3-0.6, Songbird can climb during runtime to 1.7 in just a couple of hours.

I am not changing the configuration on Songbird, although setting 'Decode' to `Pass` has the same issue. 
I have used both 

        .register_songbird_with({
                let songbird = songbird::Songbird::serenity();

                songbird.set_config(Config {
                    crypto_mode: CryptoMode::Normal,
                    decode_mode: DecodeMode::Pass,
                    preallocated_tracks: 0,
                });

                songbird
            })

and just

        .register_songbird_with()

, both having the same issue. My code for playing audio is as follows
    
    async fn play_audio(
        sound: &mut Sound,
        guild: GuildData,
        call_handler: &mut MutexGuard<'_, Call>,
        mysql_pool: MySqlPool,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        {
            let (track, track_handler) =
                create_player(sound.store_sound_source(mysql_pool.clone()).await?);
    
            let _ = track_handler.set_volume(guild.volume as f32 / 100.0);
    
            call_handler.play(track);
        }
    
        Ok(())
    }
    
...and I have defined this too for joining channels

    async fn join_channel(
        ctx: &Context,
        guild: Guild,
        channel_id: ChannelId,
    ) -> (Arc<Mutex<Call>>, JoinResult<()>) {
        let songbird = songbird::get(ctx).await.unwrap();
        let current_user = ctx.cache.current_user_id().await;
    
        let current_voice_state = guild
            .voice_states
            .get(&current_user)
            .and_then(|voice_state| voice_state.channel_id);
    
        if current_voice_state == Some(channel_id) {
            let call_opt = songbird.get(guild.id);
    
            if let Some(call) = call_opt {
                (call, Ok(()))
            } else {
                songbird.join(guild.id, channel_id).await
            }
        } else {
            songbird.join(guild.id, channel_id).await
        }
    }

...in case this is an issue in one of those rather than a general issue. I haven't got any profiling means set up right now to see where the issue is coming from much better. 
The CPU usage increases constantly during runtime. There's no 'process leak'; under `top`, all the consumption is coming from the bot and not from any left over FFMPEG sessions or anything

songbird version 0.1.4
serenity version 0.10.2
Rust version 1.49
OS: Ubuntu 18.04 headless, Linux kernel 4.15.0-135-generic

**Comment by FelixMcFelix** [comment]
Your usage looks correct to me. I can look into reproducing this or at least instrumenting as much as I can with regard to leftover threads and tasks, but I haven't really seen this climb on my end. How many servers is your bot in, is usage fairly constant?

For context, I haven't seen this in 3 weeks uptime (specs below), but my use case is small (3-4 concurrent voice connections max). The load profile there is that this bot plays one sound for every voice connection, handles every RTP/RTCP packet event, and occasionally plays a long sequence of random audio tracks.

songbird version: 0.1.4
serenity version: current
Rust version 1.50 (--release)
OS: Ubuntu 20.10, Linux kernel 5.8.0-41-generic
DigitalOcean 1 Core (1.80GHz), 1GiB RAM

**Comment by JellyWX** [comment]
Yeah, the bot is in quite a lot of voice channels at once but it shouldn't be playing audio a lot (it's a soundboard bot effectively). It's about 3k servers, so maybe up to 100 concurrent voice clients? But that's just a guess.

Audio is loaded from a number of files in /tmp, where it is fetched to from the database since I don't think I can play it directly

**Comment by FelixMcFelix** [comment]
I've hacked something together to track live threads/tasks in [this branch](https://github.com/FelixMcFelix/songbird/tree/debug-profiling). It's basic in that it just prints the counts when you call `songbird::driver::stats::global_songbird_tasks()`, but it was useful enough to patch one of the crate examples just there.

Could you try it out to see if there's a task leak that's being exposed by your environment?

**Comment by JellyWX** [comment]
it seems there's actually a second issue going on here too. The output from drivers keeps going up; currently I'm getting

    Mar 11 21:21:52 jellywx.com sfx[32285]: SONGBIRD THREAD STATS:
    Mar 11 21:21:52 jellywx.com sfx[32285]:                 CORE_TASK_COUNT: 74
    Mar 11 21:21:52 jellywx.com sfx[32285]:                 EVENT_TASK_COUNT: 74
    Mar 11 21:21:52 jellywx.com sfx[32285]:                 DISPOSAL_TASK_COUNT: 74
    Mar 11 21:21:52 jellywx.com sfx[32285]:                 MIXER_TASK_COUNT: 74
    Mar 11 21:21:52 jellywx.com sfx[32285]:                 UDP_RX_TASK_COUNT: 62
    Mar 11 21:21:52 jellywx.com sfx[32285]:                 UDP_TX_TASK_COUNT: 62
    Mar 11 21:21:52 jellywx.com sfx[32285]:                 WS_TASK_COUNT: 60

and 80% usage. But there seems to be another issue where something gets stuck and causes the 175% usage (maybe unrelated to Songbird) since the bot went unresponsive just before when that happened

**Comment by FelixMcFelix** [comment]
If you never call `Songbird::remove` then that explains the creeping task count: `leave` keeps the internal driver state intact, which can be useful for avoiding reinstalling event handlers and such. It'd be concerning if this kept rising past your active number of servers though; I'll test this locally too in case I see anything over longer time horizons.

Doesn't explain the usage though: most of the dormant tasks are waiting on MPSC messages, while the mixer threads just wake up every ~20ms to go back to sleep. Is the CPU spike and unresponsiveness accompanied by any logs of interest from songbird or serenity? I.e., failed resume, shard stopping, or similar?

**Comment by JellyWX** [comment]
I've just turned on all the logging to see what's going on with the failing stuff, and I switched out the `leave` for `remove` since I don't manually register anything additional so I don't think it matters. 
With `remove` instead of `leave`, the usage growth is certainly much slower. I've had it on a few hours now and am at around 50%.

I haven't tried that commit you have added there yet, and also the 175% hasn't tripped yet, so nothing more to add on that

**Comment by FelixMcFelix** [comment]
Glad to hear it. I can now comfortably reproduce on spawning ~10 000 `Driver`s (enough to saturate a 6-core laptop i7) -- I'm currently looking into making sure these have as close to zero cost when asleep as possible.

**Comment by JellyWX** [comment]
Brilliant, all fixed :) Thank you!
